tsne/bh_sne_src/tsne.cpp

```diff
@@ -54,10 +56,39 @@ using std::fprintf;
 using std::log;
 using std::move;
 using std::sqrt;
+using std::unique_ptr;
 using std::vector;
 
 namespace {
 
+struct TSNEState {
+ public:
+  bool step_by(int n);
+
+ private:
+  template <int NDIMS>
+  bool step_by_impl(int n);
+
+ public:
+  int N;
+  double* Y;
+  int no_dims;
+  double theta;
+  int max_iter;
+  int stop_lying_iter;
+  int mom_switch_iter;
```

Comment:
I think scattering some `const` here and there would be helpful for clarity.
```suggestion
  const int N;
  double const* Y;
  const int no_dims;
  const double theta;
  const int max_iter;
  const int stop_lying_iter;
  const int mom_switch_iter;
```

Reply:
Y is our output variable, but we do want a const pointer, so it should be `double* const Y` no?

Reply:
Yes... it's always a little confusing where the `const` goes with pointers.